h/search/client.py

```diff
@@ -23,6 +20,25 @@ def close(self):
         # to close the underlying transport directly.
         self.conn.transport.close()
 
+    @cached_property
+    def mapping_type(self):
+        """Get the name of the index's mapping type (aka. document type)."""
+        # In Elasticsearch <= 6.x our indexes have a single mapping type called
+        # "annotation". In ES >= 7.x the concept of mapping types has been
+        # removed but indexing APIs use the dummy value `_doc`.
+        # See: https://www.elastic.co/guide/en/elasticsearch/reference/6.x/removal-of-types.html
+
+        if self.server_version < Version("7.0.0"):
+            return "annotation"
+
+        return "_doc"
+
+    @cached_property
+    def server_version(self) -> Version:
+        """Get the version of the connected Elasticsearch cluster."""
+
+        return Version(self.conn.info()["version"]["number"])
```

Comment:
Used a `cached_property` as it's simpler, but also because it plays nicely with the frozen `dataclass`.

I've used a builtin `Version` object so we don't have to write any parsing code, and because it should be robust to all sorts of edge cases we might not think of. I don't think we're going to hit any of those, by why not?

Reply:
> I've used a builtin Version object so we don't have to write any parsing code, and because it should be robust to all sorts of edge cases we might not think of. 

For versions of the form X.Y.Z this will work fine. `packaging.version.Version` has a specific idea of what valid version numbers look like for things like prereleases, post-releases, build metadata etc. These might not align with ES's versioning scheme though. From what I can tell skimming the ES code for this endpoint, I think the `version.number` field will always be a "X.Y.Z" string with no other metadata.